atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/TracingKeyValueService.java

```diff
@@ -429,41 +442,107 @@ public boolean shouldTriggerCompactions() {
     }
 
-    private static CloseableTrace startLocalTrace(CharSequence operationFormat, Object... formatArguments) {
-        return CloseableTrace.startLocalTrace(SERVICE_NAME, operationFormat, formatArguments);
+    @MustBeClosed
+    private static CloseableTracer startLocalTrace(
+            @CompileTimeConstant String operation, Collection<TableReference> tableReferences) {
+        return CloseableTracer.startSpan(
+                getOperation(operation), TableReferencesTagTranslator.INSTANCE, tableReferences);
     }
 
     private static <V> ListenableFuture<V> attachDetachedSpanCompletion(
-            DetachedSpan detachedSpan, ListenableFuture<V> future, Executor tracingExecutorService) {
+            DetachedSpan detachedSpan,
+            ListenableFuture<V> future,
+            Executor tracingExecutorService,
+            Consumer<TagConsumer> tagTranslator) {
         Futures.addCallback(
                 future,
                 new FutureCallback<V>() {
                     @Override
                     public void onSuccess(V result) {
-                        detachedSpan.complete();
+                        complete();
                     }
 
                     @Override
-                    public void onFailure(Throwable throwable) {
-                        detachedSpan.complete();
+                    public void onFailure(Throwable _throwable) {
+                        complete();
+                    }
+
+                    private void complete() {
+                        detachedSpan.complete(FunctionalTagTranslator.INSTANCE, tagTranslator);
                     }
                 },
                 tracingExecutorService);
         return future;
     }
+
+    private enum TableReferenceTagTranslator implements TagTranslator<TableReference> {
+        INSTANCE;
+
+        @Override
+        public <T> void translate(TagAdapter<T> adapter, T target, TableReference data) {
+            adapter.tag(
+                    target, "table", LoggingArgs.safeTableOrPlaceholder(data).toString());
+        }
+    }
+
+    private enum TableReferencesTagTranslator implements TagTranslator<Collection<TableReference>> {
+        INSTANCE;
+
+        @Override
+        public <T> void translate(TagAdapter<T> adapter, T target, Collection<TableReference> data) {
+            adapter.tag(
+                    target, "tables", LoggingArgs.safeTablesOrPlaceholder(data).toString());
+        }
+    }
+
+    private enum NoTagTranslator implements TagTranslator<Object> {
+        INSTANCE;
+
+        @Override
+        public <T> void translate(TagAdapter<T> _adapter, T _target, Object _data) {}
+
+        @Override
+        public boolean isEmpty(Object _data) {
+            return true;
+        }
+    }
+
+    private static String getOperation(@CompileTimeConstant String operation) {
+        return SERVICE_NAME + "." + operation;
```

Comment:
ideally here too

Reply:
I'd hope the compiler is smart enough to inline these into constants?

Reply:
Done